### ISSUE TITLE
Find better exercise than Robot Simulator for the "OTP" tag

### ISSUE THREAD

**Issue by umamaistempo** [exercism/elixir#426]
The exercise _Robot Simulator_ has the OTP tag but doesn't specify that the solution must be done using processes.

Right now I am recommending the students not to use processes as this exercise doesn't require it for anything nor does it specify this need (and I used a similar exercise for code interviews and usually preferred when the candidate didn't use processes as usually when they did, it was solely to show that they know how to use an Agent and only made the code harder).

What I see people doing is starting an Agent for each robot and I don't find it adequate for several reasons:
- The robots are FSM-like machines, so what you want is a process to represent them and their actions (a GenServer), not just to hold _shared_ state*
- IF what you want is to hold state, since no one else is going to access it, you could keep that on your current scope by just assigning to a binding
- The robots don't interact in any way nor is it requested to have them all in any kind of collection, so there is no advantage in keeping them in process (actualy I think it's bad because this incentive the student to think that every different _object_ on the business model should be a different process, one of the most common conceptual mistakes from new erlang developers**)
- This makes code harder to read without any need

\* According to the own [Agent module manual](https://hexdocs.pm/elixir/Agent.html#content),
> Agents are a simple abstraction around state.
> Often in Elixir there is a need to share or store state that must be accessed from different processes or by the same process at different points in time.

\*\* TODO: find article about the missuse of processes to make separation of business entities

## TL;DR
We should probably remove the OTP tag from the Robot Simulator exercise because it is inducting students to follow bad practices

**Comment by devonestes** [comment]
While this may not be the ideal use case for a GenServer, I think this is about as good a case as we're going to find available to us in Exercism. The platform is just not designed for larger problems, the likes of which would necessitate larger, concurrent solutions like those provided by GenServer.

What this is, though, is a perfect conceptual idea (and test design) for a GenServer. When implementing this, we specifically wanted it to be solvable with either some sort of in-process state (like a struct or map), or by spinning up a process, so folks have the flexibility to learn either (or both?) ways of solving this problem, and then seeing the pros and cons of both.

This is where you mentors come in! If you see someone who just shoves state in an Agent, you can show them how to properly create an API for a GenServer, and then how to use `call` and/or `cast` in that API to implement the GenServer behaviour.

I feel pretty strongly that we should keep this tag on there. It is very likely that folks are going to reach for an Agent because many of them are new to the language. It's up to us who aren't so new to the language to offer guidance to these folks so they can learn things like the basics of using a GenServer and sending messages.

**Comment by Cohen-Carlisle** [comment]
Perhaps exercism is not the place to practice OTP at all then? I don't think it's a good idea to suggest a paradigm if the paradigm is not even needed. To the OP's point, if using OTP (which is suggested by the tags) produces worse solutions than "just" elixir, then I don't think we should suggest to students to use OTP.

Another consideration is that I've worked on 3 Elixir/Phoenix apps that run in production at my company and in none of them did we need to write code using OTP features. (Our framework/library code of course does, but that's not our source code.)

**Comment by angelikatyborska** [comment]
> If you see someone who just shoves state in an Agent, you can show them how to properly create an API for a GenServer, and then how to use call and/or cast in that API to implement the GenServer behaviour.

Why would I suggest to exchange a simple and working solution for a more complex solution?

> While this may not be the ideal use case for a GenServer, I think this is about as good a case as we're going to find available to us in Exercism.

Not true, there is Bank Account that actually requires a GenServer or Agent. There may be others, I am not familiar with all of them. If not, seeing the Bank Account, it wouldn't be that hard to add more exercises that actually require OTP.

> Another consideration is that I've worked on 3 Elixir/Phoenix apps that run in production at my company and in none of them did we need to write code using OTP features.

+1, I've worked during the past year with 5 different Elixir apps and used `GenServer` exactly once, to poll messages from AWS SQS.

**Comment by umamaistempo** [comment]
@devonestes I don't think it is a good idea to incentive someone to do something that is inadequate only for learning without a set of safety measures to ensure they won't do that anymore after understanding it.

We can make a compromise, we remove the OTP tag on this exercise and I come with another that is more suited to explain GenServers. What do you think ? :)

**Comment by NobbZ** [comment]
I've already said in the mentors slack, and I want to repeat here for public visibility:

OTP is so much more than just `GenServer`s. To be honest, in my opinion `Supervisor`s are the heart of OTP, without them, everything else are just processes…

**Comment by devonestes** [comment]
We have 92 exercises in the Elixir track at the moment, and only 3 of them include any sort of OTP (1 of which is exclusively about parallel computation, so that's a `Task.async/1` exercise). OTP is an extremely important part of the language, and I would feel like we're doing a disservice to learners by not giving them the opportunity to practice and receive feedback on OTP focused exercises. 

If there exist better fits for teaching OTP, then I'm happy to include that.

I would love to see a place where we could fit in supervision strategies, but so far we haven't been able to find one.

We are just doing the best we can in offering to learners the ability to have well rounded exposure to the important parts of the language within the boundaries of the exercises offered by Exercism. It's less than perfect for sure - like everything in computing, there are tradeoffs - but it's the best we can do for now.

**Comment by umamaistempo** [comment]
I also believe OTP centers around supervisors and genservers (as they are the bread and butter for everything; supervisors are for fault-tolerance, genservers for concurrency and the use of the OTP interface)

I believe that OTP is very important to really learning Erlang & Elixir and that we really have to give it more love on Exercism

However, I stand by my point that it's better for the student to learn about it somewhere else instead of being taught bad practices. Think that by doing this they might think they _have_ to use GenServers (or derived abstractions like Agent) everywhere (and it is true that this happens, I already met with more than one person that did this).

And as an additional point, again, it's not pointed at the exercism anywhere that the student should use processes for this exercise, besides the OTP tag and by being directed by a mentor to do it when it is not the adequate case, again, might induce them to the bad habit of using processes unnecessarily.

To finish, I think this issue was shot down too fast without the chance to be discussed by anyone, so I request it to be open again.

**Comment by devonestes** [comment]
I don't think we need to re-open the issue. Are we not in agreement that we'll try and find a new exercise to replace this one as an OTP focused exercise? Once we find that then we can remove the OTP tag from Robot Simulator.

**Comment by umamaistempo** [comment]
This issue wasn't solved yet and by closing it without solving means it's not something that is worth by itself.

Reopening means that it is an issue that should be solved and invites people to discuss

**Comment by devonestes** [comment]
Good point, it isn't solved yet, and for that reason it should be open, but I do also think we should change the title since it's now misleading. We're not just removing that tag, but we're finding a different exercise to sort of "move" that tag to. I'll update now.

**Comment by umamaistempo** [comment]
Totally agreed. Thanks :+1:

**Comment by angelikatyborska** [comment]
Two points:
- Somewhere along the way, the OTP topic ended up being removed from this exercise anyway (https://github.com/exercism/elixir/commit/bbc74115980c5d14521cf02f8287594ee23ead82)
- Topics are being deprecated in v3 (see https://github.com/exercism/elixir/issues/567). They are being replaced by "prerequisites", a list of concepts that the student needs to learn first to be able to do this exercise, otherwise the exercise will be blocked. This changes the robot simulator OTP discussion significantly because it will no longer be a suggestion, it will be a hard requirement. And there's no "OTP" concept so far anyway. We have an "agent" concept, hopefully we'll create a "genserver" concept soon, maybe "task".

Given those two points, I'm considering this discussion closed.
